### ISSUE TITLE
Bad simulation performance on `if ... { if ... { return } }`

### ISSUE THREAD

**Issue by spernsteiner** [GaloisInc/crucible#478]
`good` and `bad` do the same thing, but simulation of `good` runs fast (under a second even for N = 100) while `bad` is very slow (4 seconds for N = 12, and grows exponentially).

```C
#include <stdint.h>
#include <crucible.h>

#define N 12

int good() {
    for (int i = 0; i < N; ++i) {
        uint8_t x = crucible_uint8_t("x");
        // One symbolic branch
        if (100 < x) {
            return 1;
        }
    }
    return 0;
}

int bad() {
    for (int i = 0; i < N; ++i) {
        uint8_t x = crucible_uint8_t("x");
        // Two symbolic branches
        if (100 < x) {
            if (100 < x) {
                return 1;
            }
        }
    }
    return 0;
}

int main() {
    //return good();
    return bad();
}
```

The problem seems related to the number of symbolic branches leading up to the `return`.  For example, `if (100 < x && x < 200) return 1;` (which has 2 branches) performs like `bad`, while `if ((100 < x) & (x < 200)) return 1;` (1 branch; note use of `&`) performs like `good`.  Adding more branches, as in `x < 100 && x < 100 && x < 100`, makes performance even worse (4.5s at N = 8, growing as 3^N instead of 2^N).

Crux reports "All goals discharged through internal simplification", so the slowdown is entirely in Crucible, not the solver.

**Comment by robdockins** [comment]
You see an almost identical performance issue between the following functions:

```
#include <stdint.h>
#include <crucible.h>

#define N 12

int good2() {
  int r = 0;

    for (int i = 0; i < N; ++i) {
        uint8_t x = crucible_uint8_t("x");
        // Two symbolic branches
        if (100 < x) {
            if (100 < x) {
	      r = 1;
            }
        }
    }

    return r;
}

int bad2() {
  int r = 0;

    for (int i = 0; i < N; ++i) {
        uint8_t x = crucible_uint8_t("x");
        // Two symbolic branches
        if (100 < x) {
            if (100 < x) {
	      r = 1; break;
            }
        }
    }

    return r;
}


int main() {
  return good2();
  //return bad2();
}
```

The difference is that `good2` has a postdominator for the "if" statements inside the loop body; this means the simulator can join the paths before the next iteration and prevents exponential blow-up.  However `bad2` does an early loop exit using `break`; this means there is no postdominator inside the loop body and the simulator cannot join paths.  The result is an exponential number of paths to explore.

The moral of the story: nonlocal exits from loops are usually a bad idea for symbolic simulation. I'm not sure how to improve this situation in general.  Perhaps there is a more sophisticated analysis we can do to find more potential join points; but it's not immediately clear to me.

**Comment by robdockins** [comment]
Interestingly, running clang with `-O1` causes clang to rearrange these programs enough that that performance problems disappear.

**Comment by atomb** [comment]
Interesting! That suggests that it should be at least possible to perform efficiently on other optimization levels (even if we don't know how to at the moment).

**Comment by robdockins** [comment]
I feel like there isn't much we can do about this at the moment, but I don't want to close it because it's a pretty serious performance pitfall.  Perhaps at some point we can think of a better strategy for dealing with this relatively common code pattern.

**Comment by spernsteiner** [comment]
This seems to be fixed in latest `master`: both `bad` and `bad2` run in under a second, even with N = 200.  I tested this with clang versions 8-11.  I thought it would be necessary to enable `--opt-loop-merge` to get good performance here, but instead, adding that option makes `bad`/`bad2` time out on N = 3 and higher.

I'm closing this because I think this particular issue is now fixed, even though I'm not entirely clear on what's going on with `--opt-loop-merge`
